### PR TITLE
Don't verify stale sync responses

### PR DESCRIPTION
**Pull request by LukasGasior1**

### CODE REVIEW

File: core/src/main/java/com/radixdlt/sync/LocalSyncService.java [hunk 1]
```diff
@@ -430,10 +430,20 @@ private SyncState processSyncResponse(
     log.trace("LocalSync: Received sync response from {}", sender);
 
     if (!currentState.waitingForResponseFrom(sender)) {
-      log.warn("LocalSync: Received unexpected sync response from {}", sender);
+      log.trace("LocalSync: Received unexpected sync response from {}", sender);
+      metrics.sync().unexpectedSyncResponses().inc();
```
Comment (jakrawcz-rdx) [inline core/src/main/java/com/radixdlt/sync/LocalSyncService.java:434]: _(if you have time for such nitpicking)_ If the metric had a `sender` label, then the log would not be needed at all.

File: core/src/main/java/com/radixdlt/sync/LocalSyncService.java [hunk 2]
```diff
@@ -430,10 +430,20 @@ private SyncState processSyncResponse(
...
       return currentState;
     }
 
+    final var responseTailStateVersion =
+        syncResponse.getLedgerExtension().getTail().getLedgerHeader().getStateVersion();
+    final var currentStateVersion = currentState.getCurrentHeader().getStateVersion();
+    if (responseTailStateVersion <= currentStateVersion) {
+      log.trace("LocalSync: Received a stale sync response from {}", sender);
+      metrics.sync().staleSyncResponses().inc();
```
Comment (jakrawcz-rdx) [inline core/src/main/java/com/radixdlt/sync/LocalSyncService.java:443]: _(if you have time for such nitpicking)_ If the metric had a `sender` label, then the log would not be needed at all.

**Review by jakrawcz-rdx — APPROVED** [review]

**Comment by jakrawcz-rdx** [comment]
If I understood @dhedey correctly, he suggested to have a condition like `if response.epoch != current_epoch ...`.
Comparing versions instead seems like additional optimization (i.e. filtering out more definitely useless responses).

However, I wonder if we may still see similar issues, but caused by receiving response from _next_ epoch?
_(if that's an unrealistic risk, then comparing versions is perfect)_

File: core/src/main/java/com/radixdlt/sync/LocalSyncService.java (hunk 2 above)
Comment (malexandru-rdx) [reply core/src/main/java/com/radixdlt/sync/LocalSyncService.java:443]: I would keep the log either way.

File: core/src/main/java/com/radixdlt/sync/LocalSyncService.java (hunk 1 above)
Comment (LukasGasior1) [reply core/src/main/java/com/radixdlt/sync/LocalSyncService.java:434]: hmh, I'd prefer not to have it as a label, theoretically there isn't any upper limit of peers the node can talk to during its run lifetime

**Comment by LukasGasior1** [comment]
> However, I wonder if we may still see similar issues, but caused by receiving response from next epoch?
(if that's an unrealistic risk, then comparing versions is perfect)

I don't think it's possible. The request contains the start state version and if there's an epoch boundary in the response it must always be the tail header.
Theoretically an adversary could still send us an unsolicited "future" response, but I guess in this case it makes sense to verify it and ban that peer, if needed.

**Pull request merged**